### ISSUE TITLE
Potential memory leak for parameters [784, 200, 10]

### ISSUE THREAD

**Issue by marcdegraef** [modern-fortran/neural-fortran#24]
Hi Milan,

I tried to run the example_mnist program for a network of dimensions [784, 200, 10], with batch size of 200 and 30 epochs, learning rate of 3; I run this on Mac OS X 10.14.6 using the gfortran compiler (6.3.0) and in serial mode.

After a couple of epochs I get a system warning that my disk is full; using the activity monitor I noticed that the memory usage of the program keeps going up, eventually causing the OS to start swapping to disk until the disk is full... So I think there must be a memory leak somewhere. The total size of the input arrays and the layer arrays is only about 180Mb for this case, so it is surprising to see tens of gigs of memory allocation at runtime...

I think that the repeated calls to db_init and dw_init keep allocating new memory over and over.  I modified the train_batch routine as shown below and that appears to fix the memory leak... the overall memory usage now hovers around 200 Mb, which is as expected.

It might be possible to achieve the same with a class clean up procedure (final procedure), but I haven't tried that yet.  Let me know if you want me to do a pull request.

Thanks!
Marc.

  subroutine train_batch(self, x, y, eta)
    ! Trains a network using input data x and output data y,
    ! and learning rate eta. The learning rate is normalized
    ! with the size of the data batch.
    class(network_type), intent(in out) :: self
    real(rk), intent(in) :: x(:,:), y(:,:), eta
    type(array1d), allocatable :: db(:), db_batch(:)
    type(array2d), allocatable :: dw(:), dw_batch(:)
    integer(ik) :: i, ii, im, n, nm
    integer(ik) :: is, ie, indices(2)

    im = size(x, dim=2) ! mini-batch size
    nm = size(self % dims) ! number of layers

    ! get start and end index for mini-batch
    indices = tile_indices(im)
    is = indices(1)
    ie = indices(2)

    call db_init(db_batch, self % dims)
    call dw_init(dw_batch, self % dims)

    do concurrent(i = is:ie)
      call self % fwdprop(x(:,i))
      call self % backprop(y(:,i), dw, db)
      do concurrent(n = 1:nm)
        dw_batch(n) % array =  dw_batch(n) % array + dw(n) % array
        db_batch(n) % array =  db_batch(n) % array + db(n) % array
      end do

!----      
! new code to fix memory leak
      do ii=1,nm
       deallocate(db(ii)%array, dw(ii)%array)
      end do
      deallocate(db, dw)
!----

    end do

    if (num_images() > 1) then
      call dw_co_sum(dw_batch)
      call db_co_sum(db_batch)
    end if

    call self % update(dw_batch, db_batch, eta / im)

!----
! new code to fix memory leak
    do i=1,nm
      deallocate(db_batch(i)%array, dw_batch(i)%array)
    end do
    deallocate(db_batch, dw_batch)
!----

  end subroutine train_batch

**Comment by milancurcic** [comment]
Thanks, Marc! I'll try to reproduce and let you know what I find.

**Comment by milancurcic** [comment]
Hi Marc,

I just took a look into this in detail. I can't reproduce the issue on my laptop (gfortran-9.2.0 on Fedora 28) -- with this network size my memory usage hovers around 190 MB for the duration of the program. 

I don't know if this has to do with Mac OS or gfortran-6.3.0 or a combination of the two. Looking at the code, `db`, `dw`, `db_batch`, and `dw_batch` are allocatable arrays that are (read: should be) automatically cleaned up when the program leaves the scope of that subroutine. The fact that your added code fixes the memory leak suggests that the compiler was not doing that properly.

The best solution for this would be to update your compiler to the latest stable. Otherwise, what you already have is a solution as well.

Cheers,
milan
